File: android/src/main/kotlin/net/nativemind/flutter_llama/FlutterLlamaPlugin.kt
package net.nativemind.flutter_llama

import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * FlutterLlamaPlugin - плагин для работы с llama.cpp моделями на Android
 * 
 * Поддерживает:
 * - Загрузку GGUF моделей
 * - GPU ускорение через Vulkan/OpenCL
 * - Потоковую и обычную генерацию
 */
class FlutterLlamaPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    companion object {
        private const val TAG = "FlutterLlama"
        private const val CHANNEL_NAME = "flutter_llama"
        private const val EVENT_CHANNEL_NAME = "flutter_llama/stream"

        init {
            try {
                // Load llama.cpp libraries in correct order
                System.loadLibrary("c++_shared")
                System.loadLibrary("ggml")
                System.loadLibrary("ggml-base")
                System.loadLibrary("ggml-cpu")
                System.loadLibrary("llama")
                System.loadLibrary("flutter_llama_bridge")
                Log.d(TAG, "Native libraries loaded successfully")
            } catch (e: UnsatisfiedLinkError) {
                Log.e(TAG, "Failed to load native libraries: ${e.message}")
            }
        }
    }

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())
    
    private var modelLoaded = false
    private var modelPath: String? = null
    private var shouldStop = false

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL_NAME)
        eventChannel.setStreamHandler(this)
        
        Log.d(TAG, "Plugin attached to engine")
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "loadModel" -> loadModel(call, result)
            "generate" -> generate(call, result)
            "generateStream" -> generateStream(call, result)
            "unloadModel" -> unloadModel(result)
            "getModelInfo" -> getModelInfo(result)
            "stopGeneration" -> stopGeneration(result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        executor.shutdown()
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        shouldStop = true
    }

    // MARK: - Load Model

    private fun loadModel(call: MethodCall, result: Result) {
        executor.execute {
            try {
                val modelPath = call.argument<String>("modelPath")
                if (modelPath == null) {
                    mainHandler.post {
                        result.error("INVALID_ARGS", "Missing modelPath", null)
                    }
                    return@execute
                }

                val nThreads = call.argument<Int>("nThreads") ?: 4
                val nGpuLayers = call.argument<Int>("nGpuLayers") ?: 0
                val contextSize = call.argument<Int>("contextSize") ?: 2048
                val batchSize = call.argument<Int>("batchSize") ?: 512
                val useGpu = call.argument<Boolean>("useGpu") ?: true
                val verbose = call.argument<Boolean>("verbose") ?: false

                // Check if model file exists
                val file = File(modelPath)
                if (!file.exists()) {
                    mainHandler.post {
                        result.error("MODEL_NOT_FOUND", "Model file not found: $modelPath", null)
                    }
                    return@execute
                }

                this.modelPath = modelPath

                // Initialize model through JNI
                val success = nativeInitModel(
                    modelPath,
                    nThreads,
                    nGpuLayers,
                    contextSize,
                    batchSize,
                    useGpu,
                    verbose
                )

                modelLoaded = success

                mainHandler.post {
                    if (success) {
                        Log.d(TAG, "Model loaded: $modelPath")
                        Log.d(TAG, "GPU layers: $nGpuLayers, threads: $nThreads, context: $contextSize")
                        result.success(true)
                    } else {
                        result.error("INIT_FAILED", "Failed to initialize model", null)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading model", e)
                mainHandler.post {
                    result.error("EXCEPTION", "Error loading model: ${e.message}", null)
                }
            }
        }
    }

    // MARK: - Generate (blocking)

    private fun generate(call: MethodCall, result: Result) {
        if (!modelLoaded) {
            result.error("MODEL_NOT_LOADED", "Model not loaded", null)
            return
        }

        executor.execute {
            try {
                val prompt = call.argument<String>("prompt")
                if (prompt == null) {
                    mainHandler.post {
                        result.error("INVALID_ARGS", "Missing prompt", null)
                    }
                    return@execute
                }

                val temperature = call.argument<Double>("temperature")?.toFloat() ?: 0.8f
                val topP = call.argument<Double>("topP")?.toFloat() ?: 0.95f
                val topK = call.argument<Int>("topK") ?: 40
                val maxTokens = call.argument<Int>("maxTokens") ?: 512
                val repeatPenalty = call.argument<Double>("repeatPenalty")?.toFloat() ?: 1.1f

                shouldStop = false
                val startTime = System.currentTimeMillis()

                // Generate through JNI
                val generationResult = nativeGenerate(
                    prompt,
                    temperature,
                    topP,
                    topK,
                    maxTokens,
                    repeatPenalty
                )

                val generationTime = System.currentTimeMillis() - startTime

                mainHandler.post {
                    if (generationResult != null) {
                        val response = hashMapOf(
                            "text" to generationResult.text,
                            "tokensGenerated" to generationResult.tokensGenerated,
                            "generationTimeMs" to generationTime
                        )
                        Log.d(TAG, "Generated: ${generationResult.tokensGenerated} tokens in ${generationTime}ms")
                        result.success(response)
                    } else {
                        result.error("GENERATION_FAILED", "Failed to generate response", null)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error generating", e)
                mainHandler.post {
                    result.error("EXCEPTION", "Error generating: ${e.message}", null)
                }
            }
        }
    }

    // MARK: - Generate Stream

    private fun generateStream(call: MethodCall, result: Result) {
        if (!modelLoaded) {
            result.error("MODEL_NOT_LOADED", "Model not loaded", null)
            return
        }

        val sink = eventSink
        if (sink == null) {
            result.error("NO_EVENT_SINK", "Event channel not initialized", null)
            return
        }

        executor.execute {
            try {
                val prompt = call.argument<String>("prompt")
                if (prompt == null) {
                    mainHandler.post {
                        result.error("INVALID_ARGS", "Missing prompt", null)
                    }
                    return@execute
                }

                val temperature = call.argument<Double>("temperature")?.toFloat() ?: 0.8f
                val topP = call.argument<Double>("topP")?.toFloat() ?: 0.95f
                val topK = call.argument<Int>("topK") ?: 40
                val maxTokens = call.argument<Int>("maxTokens") ?: 512
                val repeatPenalty = call.argument<Double>("repeatPenalty")?.toFloat() ?: 1.1f

                shouldStop = false

                // Initialize streaming generation
                nativeGenerateStreamInit(prompt, temperature, topP, topK, maxTokens, repeatPenalty)

                // Stream tokens one by one
                while (!shouldStop) {
                    val token = nativeGenerateStreamNext()
                    if (token != null) {
                        mainHandler.post {
                            sink.success(token)
                        }
                    } else {
                        break
                    }
                }

                nativeGenerateStreamEnd()

                mainHandler.post {
                    sink.endOfStream()
                    result.success(null)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error in streaming generation", e)
                mainHandler.post {
                    sink.error("EXCEPTION", "Error in streaming: ${e.message}", null)
                    result.error("EXCEPTION", "Error in streaming: ${e.message}", null)
                }
            }
        }
    }

    // MARK: - Unload Model

    private fun unloadModel(result: Result) {
        if (modelLoaded) {
            nativeFreeModel()
            modelLoaded = false
            modelPath = null
            Log.d(TAG, "Model unloaded")
        }
        result.success(null)
    }

    // MARK: - Get Model Info

    private fun getModelInfo(result: Result) {
        if (!modelLoaded || modelPath == null) {
            result.success(null)
            return
        }

        try {
            val info = nativeGetModelInfo()
            if (info != null) {
                val infoMap = hashMapOf(
                    "modelPath" to modelPath!!,
                    "nParams" to info.nParams,
                    "nLayers" to info.nLayers,
                    "contextSize" to info.contextSize
                )
                result.success(infoMap)
            } else {
                result.success(null)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error getting model info", e)
            result.success(null)
        }
    }

    // MARK: - Stop Generation

    private fun stopGeneration(result: Result) {
        shouldStop = true
        nativeStopGeneration()
        result.success(null)
    }

    // MARK: - Native Methods (JNI)

    private external fun nativeInitModel(
        modelPath: String,
        nThreads: Int,
        nGpuLayers: Int,
        contextSize: Int,
        batchSize: Int,
        useGpu: Boolean,
        verbose: Boolean
    ): Boolean

    private external fun nativeGenerate(
        prompt: String,
        temperature: Float,
        topP: Float,
        topK: Int,
        maxTokens: Int,
        repeatPenalty: Float
    ): GenerationResult?

    private external fun nativeGenerateStreamInit(
        prompt: String,
        temperature: Float,
        topP: Float,
        topK: Int,
        maxTokens: Int,
        repeatPenalty: Float
    )

    private external fun nativeGenerateStreamNext(): String?

    private external fun nativeGenerateStreamEnd()

    private external fun nativeGetModelInfo(): ModelInfo?

    private external fun nativeFreeModel()

    private external fun nativeStopGeneration()

    // Data classes for JNI results
    data class GenerationResult(
        val text: String,
        val tokensGenerated: Int
    )

    data class ModelInfo(
        val nParams: Long,
        val nLayers: Int,
        val contextSize: Int
    )
}

File: ios/Classes/FlutterLlamaPlugin.swift
import Flutter
import UIKit
import Foundation

/**
 * FlutterLlamaPlugin - плагин для работы с llama.cpp моделями на iOS
 * 
 * Поддерживает:
 * - Загрузку GGUF моделей
 * - GPU ускорение через Metal
 * - Потоковую и обычную генерацию
 */
@available(iOS 13.0, *)
public class FlutterLlamaPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var modelLoaded = false
    private var modelPath: String?
    private let queue = DispatchQueue(label: "net.nativemind.flutter_llama", qos: .userInitiated)
    private var eventSink: FlutterEventSink?
    private var shouldStop = false
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "flutter_llama",
            binaryMessenger: registrar.messenger()
        )
        
        let eventChannel = FlutterEventChannel(
            name: "flutter_llama/stream",
            binaryMessenger: registrar.messenger()
        )
        
        let instance = FlutterLlamaPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)

        // Multimodal channel
        let mmChannel = FlutterMethodChannel(
            name: "flutter_llama_multimodal",
            binaryMessenger: registrar.messenger()
        )
        mmChannel.setMethodCallHandler(instance.handleMultimodal)

        NSLog("[FlutterLlama] Plugin registered (text + multimodal)")
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadModel":
            loadModel(call: call, result: result)
        case "generate":
            generate(call: call, result: result)
        case "generateStream":
            generateStream(call: call, result: result)
        case "unloadModel":
            unloadModel(result: result)
        case "getModelInfo":
            getModelInfo(result: result)
        case "stopGeneration":
            stopGeneration(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Multimodal MethodChannel Handler

    private var multimodalLoaded = false

    func handleMultimodal(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadMultimodalModel":
            loadMultimodalModel(call: call, result: result)
        case "generateMultimodal":
            generateMultimodal(call: call, result: result)
        case "unloadMultimodalModel":
            unloadMultimodalModel(result: result)
        case "getMultimodalModelInfo":
            getMultimodalModelInfo(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func loadMultimodalModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        queue.async { [weak self] in
            guard let self = self else {
                DispatchQueue.main.async { result(false) }
                return
            }
            guard self.modelLoaded else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "MODEL_NOT_LOADED", message: "Load text model first", details: nil))
                }
                return
            }
            guard let args = call.arguments as? [String: Any],
                  let mmprojPath = args["mmprojPath"] as? String else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "INVALID_ARGS", message: "Missing mmprojPath", details: nil))
                }
                return
            }

            let success = mmprojPath.withCString { cPath in
                llama_init_multimodal(cPath)
            }

            self.multimodalLoaded = success
            DispatchQueue.main.async {
                if success {
                    NSLog("[FlutterLlama] Multimodal loaded")
                    result(true)
                } else {
                    result(FlutterError(code: "MMPROJ_FAILED", message: "Failed to load mmproj", details: nil))
                }
            }
        }
    }

    private func generateMultimodal(call: FlutterMethodCall, result: @escaping FlutterResult) {
        queue.async { [weak self] in
            guard let self = self, self.modelLoaded, self.multimodalLoaded else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "NOT_READY", message: "Model or multimodal not loaded", details: nil))
                }
                return
            }
            guard let args = call.arguments as? [String: Any],
                  let prompt = args["prompt"] as? String,
                  let imageData = args["imageData"] as? FlutterStandardTypedData else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "INVALID_ARGS", message: "Missing prompt or imageData", details: nil))
                }
                return
            }

            let temperature = Float(args["temperature"] as? Double ?? 0.0)
            let maxTokens = Int32(args["maxTokens"] as? Int ?? 512)
            let startTime = Date()

            var outputBuffer = [CChar](repeating: 0, count: 16384)
            var tokensGenerated: Int32 = 0

            let bytes = imageData.data
            let success = bytes.withUnsafeBytes { rawPtr -> Bool in
                guard let imagePtr = rawPtr.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return false }
                return prompt.withCString { cPrompt in
                    llama_generate_with_image(
                        cPrompt,
                        imagePtr,
                        Int32(bytes.count),
                        temperature,
                        maxTokens,
                        &outputBuffer,
                        Int32(outputBuffer.count),
                        &tokensGenerated
                    )
                }
            }

            let generationTime = Int(Date().timeIntervalSince(startTime) * 1000)

            DispatchQueue.main.async {
                if success {
                    let text = String(cString: outputBuffer)
                    let response: [String: Any] = [
                        "text": text,
                        "tokensGenerated": tokensGenerated,
                        "generationTimeMs": generationTime,
                    ]
                    result(response)
                } else {
                    result(FlutterError(code: "GENERATE_FAILED", message: "Vision generation failed", details: nil))
                }
            }
        }
    }

    private func unloadMultimodalModel(result: @escaping FlutterResult) {
        queue.async { [weak self] in
            llama_free_multimodal()
            self?.multimodalLoaded = false
            DispatchQueue.main.async { result(nil) }
        }
    }

    private func getMultimodalModelInfo(result: @escaping FlutterResult) {
        queue.async {
            let supportsVision = llama_multimodal_supports_vision()
            let info: [String: Any] = [
                "supportsVision": supportsVision,
                "loaded": self.multimodalLoaded,
            ]
            DispatchQueue.main.async { result(info) }
        }
    }
    
    // MARK: - FlutterStreamHandler
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        shouldStop = true
        return nil
    }
    
    // MARK: - Load Model
    
    private func loadModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        queue.async { [weak self] in
            guard let self = self else { return }
            guard let args = call.arguments as? [String: Any],
                  let modelPath = args["modelPath"] as? String else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "INVALID_ARGS",
                        message: "Missing required arguments",
                        details: nil
                    ))
                }
                return
            }
            
            let nThreads = args["nThreads"] as? Int ?? 4
            let nGpuLayers = args["nGpuLayers"] as? Int ?? 0
            let contextSize = args["contextSize"] as? Int ?? 2048
            let batchSize = args["batchSize"] as? Int ?? 512
            let useGpu = args["useGpu"] as? Bool ?? true
            let verbose = args["verbose"] as? Bool ?? false
            
            // Check if model file exists
            let fileManager = FileManager.default
            guard fileManager.fileExists(atPath: modelPath) else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "MODEL_NOT_FOUND",
                        message: "Model file not found: \(modelPath)",
                        details: nil
                    ))
                }
                return
            }
            
            self.modelPath = modelPath
            
            // Initialize model through llama.cpp C++ bridge
            let success = modelPath.withCString { cPath in
                llama_init_model(
                    cPath,
                    Int32(nThreads),
                    Int32(nGpuLayers),
                    Int32(contextSize),
                    Int32(batchSize),
                    useGpu,
                    verbose
                )
            }
            
            self.modelLoaded = success
            
            DispatchQueue.main.async {
                if success {
                    NSLog("[FlutterLlama] Model loaded: \(modelPath)")
                    NSLog("[FlutterLlama] GPU layers: \(nGpuLayers), threads: \(nThreads), context: \(contextSize)")
                    result(true)
                } else {
                    result(FlutterError(
                        code: "INIT_FAILED",
                        message: "Failed to initialize model",
                        details: nil
                    ))
                }
            }
        }
    }
    
    // MARK: - Generate (blocking)
    
    private func generate(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard modelLoaded else {
            result(FlutterError(
                code: "MODEL_NOT_LOADED",
                message: "Model not loaded",
                details: nil
            ))
            return
        }
        
        queue.async { [weak self] in
            guard let self = self else { return }
            guard let args = call.arguments as? [String: Any],
                  let prompt = args["prompt"] as? String else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "INVALID_ARGS",
                        message: "Missing prompt",
                        details: nil
                    ))
                }
                return
            }
            
            let temperature = (args["temperature"] as? Double) ?? 0.8
            let topP = (args["topP"] as? Double) ?? 0.95
            let topK = (args["topK"] as? Int) ?? 40
            let maxTokens = (args["maxTokens"] as? Int) ?? 512
            let repeatPenalty = (args["repeatPenalty"] as? Double) ?? 1.1
            
            self.shouldStop = false
            let startTime = Date()
            
            // Generate through llama.cpp C++ bridge
            var outputBuffer = [CChar](repeating: 0, count: 16384)
            var tokensGenerated: Int32 = 0
            
            let success = prompt.withCString { cPrompt in
                llama_generate(
                    cPrompt,
                    Float(temperature),
                    Float(topP),
                    Int32(topK),
                    Int32(maxTokens),
                    Float(repeatPenalty),
                    &outputBuffer,
                    Int32(outputBuffer.count),
                    &tokensGenerated
                )
            }
            
            let generationTime = Int(Date().timeIntervalSince(startTime) * 1000)
            
            DispatchQueue.main.async {
                if success {
                    let responseText = String(cString: outputBuffer)
                    let response: [String: Any] = [
                        "text": responseText,
                        "tokensGenerated": Int(tokensGenerated),
                        "generationTimeMs": generationTime
                    ]
                    NSLog("[FlutterLlama] Generated: \(tokensGenerated) tokens in \(generationTime)ms")
                    result(response)
                } else {
                    result(FlutterError(
                        code: "GENERATION_FAILED",
                        message: "Failed to generate response",
                        details: nil
                    ))
                }
            }
        }
    }
    
    // MARK: - Generate Stream
    
    private func generateStream(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard modelLoaded else {
            result(FlutterError(
                code: "MODEL_NOT_LOADED",
                message: "Model not loaded",
                details: nil
            ))
            return
        }
        
        guard let eventSink = self.eventSink else {
            result(FlutterError(
                code: "NO_EVENT_SINK",
                message: "Event channel not initialized",
                details: nil
            ))
            return
        }
        
        queue.async { [weak self] in
            guard let self = self else { return }
            guard let args = call.arguments as? [String: Any],
                  let prompt = args["prompt"] as? String else {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "INVALID_ARGS",
                        message: "Missing prompt",
                        details: nil
                    ))
                }
                return
            }
            
            let temperature = (args["temperature"] as? Double) ?? 0.8
            let topP = (args["topP"] as? Double) ?? 0.95
            let topK = (args["topK"] as? Int) ?? 40
            let maxTokens = (args["maxTokens"] as? Int) ?? 512
            let repeatPenalty = (args["repeatPenalty"] as? Double) ?? 1.1
            
            self.shouldStop = false
            
            // Initialize streaming generation
            prompt.withCString { cPrompt in
                llama_generate_stream_init(
                    cPrompt,
                    Float(temperature),
                    Float(topP),
                    Int32(topK),
                    Int32(maxTokens),
                    Float(repeatPenalty)
                )
            }
            
            // Stream tokens one by one
            var tokenBuffer = [CChar](repeating: 0, count: 256)
            while !self.shouldStop {
                let hasMore = llama_generate_stream_next(&tokenBuffer, Int32(tokenBuffer.count))
                
                if hasMore {
                    let token = String(cString: tokenBuffer)
                    DispatchQueue.main.async {
                        eventSink(token)
                    }
                } else {
                    break
                }
            }
            
            llama_generate_stream_end()
            
            DispatchQueue.main.async {
                eventSink(FlutterEndOfEventStream)
                result(nil)
            }
        }
    }
    
    // MARK: - Unload Model
    
    private func unloadModel(result: @escaping FlutterResult) {
        if modelLoaded {
            llama_cpp_bridge_free_model()
            modelLoaded = false
            modelPath = nil
            NSLog("[FlutterLlama] Model unloaded")
        }
        result(nil)
    }
    
    // MARK: - Get Model Info
    
    private func getModelInfo(result: @escaping FlutterResult) {
        guard modelLoaded, let modelPath = modelPath else {
            result(nil)
            return
        }
        
        var nParams: Int64 = 0
        var nLayers: Int32 = 0
        var contextSize: Int32 = 0
        
        llama_get_model_info(&nParams, &nLayers, &contextSize)
        
        let info: [String: Any] = [
            "modelPath": modelPath,
            "nParams": nParams,
            "nLayers": nLayers,
            "contextSize": contextSize
        ]
        
        result(info)
    }
    
    // MARK: - Stop Generation
    
    private func stopGeneration(result: @escaping FlutterResult) {
        shouldStop = true
        llama_stop_generation()
        result(nil)
    }
}

// MARK: - C++ Bridge Function Declarations

// These functions will be implemented in llama_cpp_bridge.cpp
@_silgen_name("llama_init_model")
func llama_init_model(
    _ modelPath: UnsafePointer<CChar>,
    _ nThreads: Int32,
    _ nGpuLayers: Int32,
    _ contextSize: Int32,
    _ batchSize: Int32,
    _ useGpu: Bool,
    _ verbose: Bool
) -> Bool

@_silgen_name("llama_generate")
func llama_generate(
    _ prompt: UnsafePointer<CChar>,
    _ temperature: Float,
    _ topP: Float,
    _ topK: Int32,
    _ maxTokens: Int32,
    _ repeatPenalty: Float,
    _ output: UnsafeMutablePointer<CChar>,
    _ outputSize: Int32,
    _ tokensGenerated: UnsafeMutablePointer<Int32>
) -> Bool

@_silgen_name("llama_generate_stream_init")
func llama_generate_stream_init(
    _ prompt: UnsafePointer<CChar>,
    _ temperature: Float,
    _ topP: Float,
    _ topK: Int32,
    _ maxTokens: Int32,
    _ repeatPenalty: Float
)

@_silgen_name("llama_generate_stream_next")
func llama_generate_stream_next(
    _ output: UnsafeMutablePointer<CChar>,
    _ outputSize: Int32
) -> Bool

@_silgen_name("llama_generate_stream_end")
func llama_generate_stream_end()

@_silgen_name("llama_get_model_info")
func llama_get_model_info(
    _ nParams: UnsafeMutablePointer<Int64>,
    _ nLayers: UnsafeMutablePointer<Int32>,
    _ contextSize: UnsafeMutablePointer<Int32>
)

@_silgen_name("llama_free_model")
func llama_cpp_bridge_free_model()

@_silgen_name("llama_stop_generation")
func llama_stop_generation()

// Multimodal (Vision) bridge functions
@_silgen_name("llama_init_multimodal")
func llama_init_multimodal(_ mmprojPath: UnsafePointer<CChar>) -> Bool

@_silgen_name("llama_generate_with_image")
func llama_generate_with_image(
    _ prompt: UnsafePointer<CChar>,
    _ imageData: UnsafePointer<UInt8>,
    _ imageLen: Int32,
    _ temperature: Float,
    _ maxTokens: Int32,
    _ output: UnsafeMutablePointer<CChar>,
    _ outputSize: Int32,
    _ tokensGenerated: UnsafeMutablePointer<Int32>
) -> Bool

@_silgen_name("llama_free_multimodal")
func llama_free_multimodal()

@_silgen_name("llama_multimodal_supports_vision")
func llama_multimodal_supports_vision() -> Bool
